pkg/receive/handler.go

```diff
@@ -747,82 +761,49 @@ func (h *Handler) fanoutForward(pctx context.Context, tenant string, wreqs map[e
 	// This is needed if context is canceled or if we reached success of fail quorum faster.
 	defer func() {
 		go func() {
-			for err := range ec {
-				if err != nil {
-					level.Debug(tLogger).Log("msg", "request failed, but not needed to achieve quorum", "err", err)
+			for wresp := range ec {
+				if wresp.err != nil {
+					level.Debug(tLogger).Log("msg", "request failed, but not needed to achieve quorum", "err", wresp.err)
 				}
 			}
 		}()
 	}()
 
-	var success int
+	quorum := h.writeQuorum()
+	failureThreshold := h.options.ReplicationFactor - uint64(quorum)
+	successes := make([]int, numSeries)
+	seriesErrs := make([]errutil.MultiError, numSeries)
 	for {
 		select {
 		case <-fctx.Done():
 			return fctx.Err()
-		case err, more := <-ec:
+		case wresp, more := <-ec:
 			if !more {
+				for _, rerr := range seriesErrs {
+					if seriesReplicated {
+						errs.Add(rerr.Err())
+					} else if uint64(len(rerr)) >= failureThreshold {
+						cause := determineWriteErrorCause(rerr.Err(), quorum)
```

Comment:
I think we'll also have to change how we determine the HTTP error we return to client, when this `cause` error bubbles up back to `handleRequest`. Right now, we'll return error that occurs the most or the original multi error, since we use threshold `1`. But this might be incorrect, as if a `cause` error for any individual series replications will be server error, we have to retry the whole request. I think solution would be:

- Return server error, if **any** of the `cause` errors is an unknown error / unavailable / not ready (cases when we have to retry). Tricky but less important part here might be exactly which error to return if we have a mixed bag of server errors - the behavior of client should be same though regardless of the error message we decide to return.
- Otherwise we should only have conflict errors and can return conflict

Reply:
Another thing we should be mindful of here is when `cause` will return the original multi-error (and same actually above for the `if` branch), we are putting a multi-error inside of the `errs` multi-error, which can lead to erroneous `5xx` as described in https://github.com/thanos-io/thanos/issues/5407#issuecomment-1148950914 .

Reply:
Yes you are correct. However I wonder if we already have this issue in main because we calculate the top-level cause the same way using `threshold=1`. So if we have 2 batches with conflict and 1 batch with server error, we will return conflict to the user and not retry the request.

In any case, I would also prefer to solve this problem now since it can lead to data loss.

Reply:
One thing I am not sure about is what should the error code be when we try to replicate a series and we get one success, one server error and one client error. Right now I believe we return client-error, but if we change the rules, we would return a server error. It also means that in case of 2 conflicts (samples already exist in TSDB) and 1 server error, we would still return a server error even though that might not be necessary. 

Maybe for replicating an individual series we can treat client-errors as success and only return 5xx when two replicas fail. For the overall response, we can return 5xx if any series has a 5xx.

Reply:
Yes, I believe we basically have to treat conflict as if we have 'success'. It's just important to return the correct status to the upstream so if we have any conflicts in the replication, we'll want to return this to the client. Otherwise 5xx and OK should be clear (5xx if any series fails quorum; OK if no failed quorums or conflicts).

Reply:
Makes sense. I think the `MultiError` and `determineWriteErrorCause` are not good abstractions for this. The `determineWriteErrorCause` function is overloaded and tries to determine the error for both cases. Because of this, I added a `writeErrors` type which has two separate methods, `Cause()` and `replicationErr()`. `Cause` is used to determine the error across multiple series, while `replicationErr` is used for determining the error of replicating a single series.